drivers/video/video_buffer.c

```diff
@@ -0,0 +1,229 @@
+}
+
+void video_buffer_release(struct video_buffer *vbuf)
+{
+	struct mem_block *block = NULL;
+	int i;
+
+	__ASSERT_NO_MSG(vbuf != NULL);
+
+	/* vbuf to block */
+	for (i = 0; i < ARRAY_SIZE(video_block); i++) {
+		if (video_block[i].data == vbuf->buffer) {
+			block = &video_block[i];
+			break;
+		}
+	}
+
+	vbuf->buffer = NULL;
+	if (block) {
+		VIDEO_COMMON_FREE(block->data);
+	}
+}
+
+/* To be completed */
+int video_request_buffers(uint8_t count, size_t size, enum video_buf_type type,
+			  enum video_buf_memory memory)
+{
+	struct video_buffer *buffer;
+
+	/* TODO: Input request vs Output request for m2m devices */
+	if (count > CONFIG_VIDEO_BUFFER_POOL_NUM_MAX) {
+		return -EINVAL;
+	}
+
+	/* TODO: Buffer request is not cumulative. For each new request, free all the buffers
+	 * that are previously requested
+	 */
+
+	/* Allocate buffers */
+	for (uint8_t i = 0; i < count; i++) {
+		if (memory == VIDEO_MEMORY_VIDEO) {
+			buffer = video_buffer_aligned_alloc(size, CONFIG_VIDEO_BUFFER_POOL_ALIGN,
+							    K_FOREVER);
+		} else {
+			buffer = &video_buf[i];
+		}
+
+		buffer->type = type;
+		buffer->index = i;
+		buffer->size = size;
+		buffer->bytesused = 0;
+		buffer->state = VIDEO_BUF_STATE_DONE;
+	}
+
+	return 0;
+}
+
+RTIO_DEFINE(rtio, CONFIG_VIDEO_BUFFER_POOL_NUM_MAX, CONFIG_VIDEO_BUFFER_POOL_NUM_MAX);
+
+int video_enqueue(const struct device *dev, struct video_buffer *buf)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(dev != NULL);
+
+	const struct video_driver_api *api = (const struct video_driver_api *)dev->api;
+	if (api->enqueue == NULL) {
+		return -ENOSYS;
+	}
+
+	if (video_buf[buf->index].type != buf->type ||
+	    video_buf[buf->index].memory != buf->memory ||
+	    video_buf[buf->index].state != VIDEO_BUF_STATE_DONE) {
+		return -EINVAL;
+	}
+
+	if (buf->memory == VIDEO_MEMORY_USER) {
+		if (buf->size != video_buf[buf->index].size) {
+			return -EINVAL;
+		}
+
+		video_buf[buf->index].buffer = buf->buffer;
+	}
+
+	ret = api->enqueue(dev, &video_buf[buf->index]);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* RTIO submission */
+	struct rtio_iodev *ri = video_find_iodev(dev);
+	struct rtio_sqe *sqe = rtio_sqe_acquire(&rtio);
+
+	__ASSERT_NO_MSG(ri != NULL);
+	__ASSERT_NO_MSG(sqe != NULL);
+
+	rtio_sqe_prep_read(sqe, ri, RTIO_PRIO_NORM, video_buf[buf->index].buffer,
+			   video_buf[buf->index].size, &video_buf[buf->index]);
+
+	sqe->flags |= RTIO_SQE_MULTISHOT;
+
+	/* Do not wait for complete */
+	rtio_submit(&rtio, 0);
+
+	video_buf[buf->index].state = VIDEO_BUF_STATE_QUEUED;
+
+	return 0;
+}
+
+int video_dequeue(struct video_buffer **buf)
+{
+	struct rtio_cqe *cqe = rtio_cqe_consume_block(&rtio);
+	*buf = cqe->userdata;
+
+	if (cqe->result < 0) {
+		LOG_ERR("I/O operation failed");
+		return cqe->result;
+	}
+
+	return 0;
+}
+
+void video_release_buf()
+{
+	/* Buffer will be re-queued thanks to RTIO_SQE_MULTISHOT */
+	rtio_cqe_release(&rtio, rtio_cqe_consume_block(&rtio));
+}
```

Comment:
There is `video_buffer_release()` and `video_release_buf()`. To avoid confusion, how about `video_buffer_done()` or other name?